File: backend/routers/sessions.py
from __future__ import annotations

import json
import logging
from dataclasses import asdict
from datetime import UTC, datetime

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlmodel import Session as DBSession
from sqlmodel import select

from config import settings
from db import get_db
from models.tables import Document, Message, SessionDocument
from models.tables import Session as SessionModel
from services.model_catalog import is_valid_provider, known_providers
from services.rag_config import RAGConfig

router = APIRouter(prefix="/api/sessions", tags=["sessions"])
logger = logging.getLogger("chatpdf.sessions")


def _parse_sources(raw: str | None) -> list[dict] | None:
    """Defensively parse the stored sources JSON — a malformed blob must not 500
    a session read."""
    if not raw:
        return None
    try:
        data = json.loads(raw)
        return data if isinstance(data, list) else None
    except (ValueError, TypeError):
        logger.warning("could not parse message.sources JSON; returning null")
        return None


# ── Schemas ───────────────────────────────────────────────────────────────────

class SessionCreate(BaseModel):
    doc_ids: list[str]
    provider: str
    model: str
    title: str | None = None
    rag_config: dict | None = None


class SessionUpdate(BaseModel):
    title: str


class DocumentSnippet(BaseModel):
    doc_id: str
    name: str
    status: str


class MessageOut(BaseModel):
    role: str
    content: str
    sources: list[dict] | None = None
    created_at: str


class SessionOut(BaseModel):
    session_id: str
    title: str
    documents: list[DocumentSnippet]
    rag_config: dict
    provider: str
    model: str
    created_at: str


class SessionDetailOut(SessionOut):
    messages: list[MessageOut]


def _msg_out(m: Message) -> MessageOut:
    return MessageOut(
        role=m.role,
        content=m.content,
        sources=_parse_sources(m.sources),
        created_at=m.created_at.isoformat(),
    )


def _session_docs(session_id: str, db: DBSession) -> list[Document]:
    rows = db.exec(
        select(Document)
        .join(SessionDocument, SessionDocument.document_id == Document.id)
        .where(SessionDocument.session_id == session_id)
        .order_by(SessionDocument.added_at)
    ).all()
    return list(rows)


def _session_out(s: SessionModel, docs: list[Document]) -> SessionOut:
    return SessionOut(
        session_id=s.id,
        title=s.title,
        documents=[DocumentSnippet(doc_id=d.id, name=d.name, status=d.status) for d in docs],
        # Normalize through RAGConfig so the response is always a complete, valid
        # config (defaults filled, unknown keys dropped) — never a raw/legacy blob.
        rag_config=asdict(RAGConfig.from_json(s.rag_config)),
        provider=s.provider,
        model=s.model,
        created_at=s.created_at.isoformat(),
    )


# ── Routes ────────────────────────────────────────────────────────────────────

@router.post("", status_code=201, response_model=SessionOut)
def create_session(body: SessionCreate, db: DBSession = Depends(get_db)):
    if not is_valid_provider(body.provider):
        raise HTTPException(
            status_code=422,
            detail=f"Unknown provider '{body.provider}'. Choose one of: {', '.join(known_providers())}",
        )
    if len(body.doc_ids) > settings.max_docs_per_session:
        raise HTTPException(
            status_code=422,
            detail=f"Too many documents: limit is {settings.max_docs_per_session} per session.",
        )
    for doc_id in body.doc_ids:
        if not db.get(Document, doc_id):
            raise HTTPException(status_code=404, detail=f"Document not found: {doc_id}")
    session = SessionModel(
        title=body.title or "New Chat",
        rag_config=json.dumps(body.rag_config or {}),
        provider=body.provider,
        model=body.model,
    )
    db.add(session)
    db.commit()
    db.refresh(session)
    for doc_id in body.doc_ids:
        db.add(SessionDocument(session_id=session.id, document_id=doc_id))
    db.commit()
    return _session_out(session, _session_docs(session.id, db))


@router.get("", response_model=list[SessionOut])
def list_sessions(db: DBSession = Depends(get_db)):
    rows = db.exec(select(SessionModel).order_by(SessionModel.created_at.desc())).all()
    return [_session_out(s, _session_docs(s.id, db)) for s in rows]


@router.delete("", status_code=204)
def delete_all_sessions(db: DBSession = Depends(get_db)):
    """Delete every session (messages + document links cascade)."""
    for s in db.exec(select(SessionModel)).all():
        db.delete(s)
    db.commit()


@router.get("/{session_id}", response_model=SessionDetailOut)
def get_session(session_id: str, db: DBSession = Depends(get_db)):
    s = db.get(SessionModel, session_id)
    if not s:
        raise HTTPException(status_code=404, detail="Session not found")
    msgs = db.exec(
        select(Message).where(Message.session_id == session_id).order_by(Message.created_at)
    ).all()
    base = _session_out(s, _session_docs(session_id, db))
    return SessionDetailOut(**base.model_dump(), messages=[_msg_out(m) for m in msgs])


@router.patch("/{session_id}", response_model=SessionOut)
def rename_session(session_id: str, body: SessionUpdate, db: DBSession = Depends(get_db)):
    s = db.get(SessionModel, session_id)
    if not s:
        raise HTTPException(status_code=404, detail="Session not found")
    s.title = body.title
    s.updated_at = datetime.now(UTC)
    db.add(s)
    db.commit()
    db.refresh(s)
    return _session_out(s, _session_docs(session_id, db))


@router.delete("/{session_id}", status_code=204)
def delete_session(session_id: str, db: DBSession = Depends(get_db)):
    s = db.get(SessionModel, session_id)
    if not s:
        raise HTTPException(status_code=404, detail="Session not found")
    db.delete(s)
    db.commit()

File: frontend/src/stores/sessions.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '@/lib/api'

export interface MessageOut {
  role: 'user' | 'assistant'
  content: string
  sources: Array<{ doc_name: string; page?: number | null; chunk_preview: string; score: number }> | null
  created_at: string
}

export interface DocumentSnippet {
  doc_id: string
  name: string
  status: string
}

export interface Session {
  session_id: string
  title: string
  documents: DocumentSnippet[]
  rag_config: Record<string, unknown>
  provider: string
  model: string
  created_at: string
}

export interface SessionDetail extends Session {
  messages: MessageOut[]
}

export const useSessionsStore = defineStore('sessions', () => {
  const sessions = ref<Session[]>([])
  const activeSession = ref<SessionDetail | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)
  const models = ref<Record<string, string[]>>({})

  async function fetchModels(): Promise<void> {
    try {
      models.value = await api.get<Record<string, string[]>>('/api/models')
    } catch {
      // Non-fatal: the UI falls back to its built-in defaults.
    }
  }

  async function fetchSessions(): Promise<void> {
    loading.value = true
    error.value = null
    try {
      sessions.value = await api.get<Session[]>('/api/sessions')
    } catch (e) {
      error.value = e instanceof Error ? e.message : String(e)
    } finally {
      loading.value = false
    }
  }

  async function fetchSession(id: string): Promise<SessionDetail> {
    const detail = await api.get<SessionDetail>(`/api/sessions/${id}`)
    activeSession.value = detail
    return detail
  }

  async function createSession(payload: {
    doc_ids: string[]
    provider: string
    model: string
    title?: string
    rag_config?: Record<string, unknown>
  }): Promise<Session> {
    const sess = await api.postJson<Session>('/api/sessions', payload)
    sessions.value.unshift(sess)
    return sess
  }

  async function renameSession(id: string, title: string): Promise<void> {
    const updated = await api.patchJson<Session>(`/api/sessions/${id}`, { title })
    const idx = sessions.value.findIndex(s => s.session_id === id)
    if (idx !== -1) sessions.value[idx] = updated
    if (activeSession.value?.session_id === id) {
      activeSession.value.title = updated.title
    }
  }

  async function deleteSession(id: string): Promise<void> {
    await api.del(`/api/sessions/${id}`)
    sessions.value = sessions.value.filter(s => s.session_id !== id)
    if (activeSession.value?.session_id === id) activeSession.value = null
  }

  async function deleteAllSessions(): Promise<void> {
    await api.del('/api/sessions')
    sessions.value = []
    activeSession.value = null
  }

  return {
    sessions, activeSession, loading, error, models,
    fetchModels, fetchSessions, fetchSession, createSession, renameSession, deleteSession, deleteAllSessions,
  }
})
